### ISSUE TITLE
Middleware returning empty 200 for old build ID instead of 404

### ISSUE THREAD

**Issue by ben-propflo** [vercel/next.js#57207]
### Link to the code that reproduces this issue

https://github.com/ben-propflo/nextjs-middleware-deployment-repo

### To Reproduce

1. Clone the repro
2. Run yarn build && yarn start
3. Load http://localhost:3000/ssr and keep it open
4. Stop Next.js
5. Re-run yarn build && yarn start
6. Click the SSR2 link, no SSR prop data is returned. A request is made to "http://localhost:3000/_next/data/build-id/ssr.json" which returns a 200 and an empty object. This will generally cause a client side exception as when rendering the page, props won't exist.

If you then delete middleware.ts and repeat the steps above, you'll notice that the request to _next/data/ will return a 404 which causes a new fetch to http://localhost:3000/ssr2 so everything works as expected.

### Current vs. Expected behavior

Following the steps from the previous section, when using middleware I'd expect the same behaviour as when not using middleware. That is a 404 is returned so that the client re-fetches the page.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Fri Oct  6 19:02:35 UTC 2023
Binaries:
  Node: 18.18.0
  npm: 8.18.0
  Yarn: 1.22.19
  pnpm: N/A
Relevant Packages:
  next: 13.5.7-canary.14
  eslint-config-next: 13.5.6
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Data fetching (gS(S)P, getInitialProps), Middleware / Edge (API routes, runtime), Routing (next/router, next/navigation, next/link)

### Additional context

I've gone back and tried previous versions of Next.js on my linked repro, the issue first starts appearing in [v13.4.13-canary.0](https://github.com/vercel/next.js/releases/tag/v13.4.13-canary.0) (it works fine on 13.4.12)

**Comment by EmilioAiolfi** [comment]
Same happens with version 13.5.3
![Screenshot 2023-10-23 at 10 57 21](https://github.com/vercel/next.js/assets/999120/77d03ac4-438c-49d4-871c-2339dbcdd726)

**Comment by ben-propflo** [comment]
I've re-tested on [v14.0.0](https://github.com/vercel/next.js/releases/tag/v14.0.0) and that still has the same broken behaviour.

**Comment by ben-propflo** [comment]
I've updated the repo to [14.0.3-canary.9](https://github.com/vercel/next.js/releases/tag/v14.0.3-canary.9) and the same broken behaviour **still exists**.

**Comment by ben-propflo** [comment]
Did a bit more digging around, I think this is the problem commit:[routing](https://github.com/vercel/next.js/commit/1398de9977b89c9c4717d26e213b52bc63ccbe7e)

There are some changes in packages/next/src/server/next-server.ts, specifically L1228, where the buildId check disappears. The commit is rather massive so it might have been moved elsewhere?

**Comment by ben-propflo** [comment]
Updated the repro repo, the issue is still reproducible on [14.1.1-canary.7](https://github.com/vercel/next.js/releases/tag/v14.1.1-canary.7)

**Comment by sabreu-teknorix** [comment]
Issue is still reproducible on [14.1.1-canary.27](https://github.com/vercel/next.js/releases/tag/v14.1.1-canary.27)

**Comment by ben-propflo** [comment]
This is resolved in [v14.1.1-canary.37](https://github.com/vercel/next.js/releases/tag/v14.1.1-canary.37) :rocket:
